[Feat] Detector implementation for Azure API Management Direct Management Key

abmussani opened the pull request:
### Description:
This PR implements a new Detector for Azure Direct management API key. This detector looks for direct manage url and keys (Primary or Secondary), generates a access token with expiry of 5 seconds (to avoid leakage) and verify the credentials based on generated values.

Test Output:
```
$ go test -tags=detectors ./pkg/detectors/azuredirectmanagementkey/
ok  	github.com/trufflesecurity/trufflehog/v3/pkg/detectors/azuredirectmanagementkey	6.881s
```

### Checklist:
* [ ] Tests passing (`make test-community`)?
* [x] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/welcome/install/#local-installation))?

abmussani committed:
azuredirectmanagementapikey detector implementation and tests

abmussani committed:
add reference url

abmussani committed:
use sanehttpclient

abmussani committed:
better regex

abmussani committed:
add prefix regex for key pattern

abmussani committed:
some refactoring

abmussani committed:
cleanup some unnecessary logic for expiry formatting.
code refactoring, remove unnecessary code.
updated the description of detector.
reduce the expiry to 5 seconds.

rgmz commented on pkg/detectors/azuredirectmanagementkey/azuredirectmanagementkey.go:
```diff
@@ -0,0 +1,142 @@
+			s1 := detectors.Result{
+				DetectorType: detectorspb.DetectorType_AzureDirectManagementKey,
+				Raw:          []byte(baseUrl),
+				RawV2:        []byte(baseUrl + primaryKey),
```
Can we start making `RawV2` structured in new detectors? https://github.com/trufflesecurity/trufflehog/pull/3634#discussion_r1851024698

abmussani replied:
@rgmz I am also not in favor of using `RawV2` as structured secret. As @ahrav mentioned that it is tightly coupled with Enterprise and (in future) if we need to restructure it, due to any reason, We wont be able to do it. 

My suggestion is to have another `Map` for this purpose.

rgmz replied:
> ... it is tightly coupled with Enterprise and (in future) if we need to restructure it, due to any reason, We wont be able to do it.

That is the case regardless. Making it structured at least makes it parseable and human-readable, unlike doing a simple concatenation.

rgmz commented on pkg/detectors/azuredirectmanagementkey/azuredirectmanagementkey.go:
```diff
@@ -0,0 +1,142 @@
+var (
+	defaultClient = common.SaneHttpClient()
+	urlPat        = regexp.MustCompile(`https://([a-z0-9][a-z0-9-]{0,48}[a-z0-9])\.management\.azure-api\.net`)         // https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.APIM.Name/
+	primaryKeyPat = regexp.MustCompile(detectors.PrefixRegex([]string{"azure"}) + `\b([a-zA-Z0-9+\/-]{86,88}\b={0,2})`) // Base64-encoded primary key
```
Why `-` at the end?

abmussani replied:
@rgmz Initially I copied that regex from another detector, [Azure Storage](https://github.com/trufflesecurity/trufflehog/blob/8765cc6bcb9e5a81105b3d5a131b8e6fbd6fd34b/pkg/detectors/azure_storage/storage.go#L32). After your comment, I generated 20 times a key and confidently say that all of them ended with `[a-zA-Z0-9]==`. Updated the regex.

abmussani committed:
improvise the regex to increase confidence level.
refactor variable name for better understanding.

rgmz commented on pkg/detectors/azuredirectmanagementkey/azuredirectmanagementkey.go:
```diff
@@ -0,0 +1,142 @@
+var (
+	defaultClient = common.SaneHttpClient()
+	urlPat        = regexp.MustCompile(`https://([a-z0-9][a-z0-9-]{0,48}[a-z0-9])\.management\.azure-api\.net`)                                          // https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.APIM.Name/
+	keyPat        = regexp.MustCompile(detectors.PrefixRegex([]string{"azure", ".management.azure-api.net"}) + `\b([a-zA-Z0-9+\/]{83,85}[a-zA-Z0-9]==)`) // pattern for both Primary and secondary key
```
`\b` isn't valid when beside `/` or `+`.

ahrav replied:
This has bitten me so many times... 😭

abmussani replied:
Updated the regex but [AzureContainerRegistry](https://github.com/trufflesecurity/trufflehog/blob/8765cc6bcb9e5a81105b3d5a131b8e6fbd6fd34b/pkg/detectors/azurecontainerregistry/azurecontainerregistry.go#L33) Detector has the similar issue. right ?

abmussani committed:
remove \b from regex

abmussani committed:
implemented CustomFalsePositiveChecker interface.
implemented host caching if host is not reachable to avoid unnecessary extra verification calls.

abmussani committed:
add delimiter to raw v2

kashifkhan0771 commented on pkg/detectors/azuredirectmanagementkey/azuredirectmanagementkey.go:
```diff
@@ -0,0 +1,163 @@
+	return "Azure API Management provides a direct management REST API for performing operations on selected entities, such as users, groups, products, and subscriptions."
+}
+
+func (s Scanner) IsFalsePositive(_ detectors.Result) (bool, string) {
```
If we want to skip some specific keyword can we implement it as [this](https://github.com/trufflesecurity/trufflehog/blob/main/pkg/detectors/slackwebhook/slackwebhook.go#L117)

kashifkhan0771 reviewed (approved):
Just one comment related to False Positive Implementation. Rest LGTM!

zricethezav commented:
@abmussani gotta fix conflicts here too

abmussani committed:
Merge branch 'main' into azuredirectmanagementapikey-detector-impl

abmussani committed:
make protos

abmussani committed:
Merge branch 'main' into azuredirectmanagementapikey-detector-impl

abmussani committed:
Merge branch 'main' into azuredirectmanagementapikey-detector-impl
